ipcc-key-value/src/ioctl.rs

```diff
@@ -0,0 +1,133 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2023 Oxide Computer Company
+
+//! IPCC `ioctl` interface.
+//!
+//! This module is tightly-coupled to the host OS image, and should only be used
+//! by services bundled with it (e.g., sled-agent and installinator).
+
+use crate::InstallinatorImageId;
+use crate::InstallinatorImageIdError;
+use crate::IpccKeyLookupError;
+use crate::PingError;
+use crate::Pong;
+use libc::c_int;
+use std::fs::File;
+use std::io;
+use std::os::unix::prelude::AsRawFd;
+
+pub struct Ipcc {
+    file: File,
+}
+
+impl Ipcc {
+    pub fn open() -> io::Result<Self> {
+        let file = File::options().read(true).write(true).open(IPCC_DEV)?;
+        Ok(Self { file })
+    }
+
+    pub fn ping(&self) -> Result<Pong, PingError> {
+        const EXPECTED_REPLY: &[u8] = b"pong";
+
+        let mut buf = [0; EXPECTED_REPLY.len()];
+        let n = self.key_lookup(IpccKey::Ping, &mut buf)?;
+        let buf = &buf[..n];
+
+        if buf == EXPECTED_REPLY {
+            Ok(Pong)
+        } else {
+            Err(PingError::UnexpectedReply(buf.to_vec()))
+        }
+    }
+
+    pub fn installinator_image_id(
+        &self,
+    ) -> Result<InstallinatorImageId, InstallinatorImageIdError> {
+        let mut buf = [0; InstallinatorImageId::CBOR_SERIALIZED_SIZE];
```

Comment:
Mostly just thinking aloud: if we ever add more data to this, we'll be able to update the installinator image to handle the additional data so it's OK to use the exact size here.

Reply:
Yeah, I was doing the same thinking. We _could_ pass a bigger buffer here, but if we got back more data than expected it's not going to deserialize anyway.